Refactor the cake so SymbolTable does not depend on Global

gkossakowski opened the pull request:
_The description of PR is copied from 364fcd9. All other commits are mostly supporting that big change._

This is rather large commit so I'll first explain the motivation behind it and then go through all changes in detail explaining the choices I made.

The motivation behind this refactoring was to make SymbolTable unit testable. I wanted a lightweight way of initializing SymbolTable and then writing unit tests for subtyping algorithm, various functionality related to Symbols, etc. All of that should be possible by precisely controlling what we test, e.g., create types and symbols by hand and not have them defined in source code as we normally do in partest (functional) tests.

The other motivation was to reduce and clarify dependencies we have in the compiler. Explicit dependencies lead to cleaner design. Also, explicit and reduced dependencies help incremental compilation which is a big problem for us in compiler's code base at the moment.

One of the challenges I faced during that refactoring was cyclic dependency between Platform and SymbolLoaders.

Platform depended on `SymbolLoaders.SymbolLoader` because it would define a root loader. SymbolLoaders depended on Platform for numerous reasons like deferring decision how to load a given symbol based on some Platform-specific hooks.

I decided to break that cycle by removing methods related to symbol loading from Platform interface. One could argue, that better fix would be to make SymbolLoaders to not depend on Platform (backend) concept but that would be much bigger refactoring. Also, we have a new concept for dealing with symbol loading: Mirrors.

For those reasons both `newClassLoader` and `rootLoader` were dropped from Platform interface.

Note that JavaPlatform still depends on Global so it can access phases defined in Global to implement `platformPhases` method.

Both GenICode and BCodeBodyBuilder have some Platform specific logic that requires casting because pattern matcher doesn't narrow types to give them a proper refinement. Check the changes for details.

Some logging utilities has been moved from Global to SymbolTable because they are accessed by SymbolTable. Since Global inherits from SymbolTable this should be a source compatible change.

The SymbolLoaders has dependency on `compileLate` method defined in Global. The purpose behind `compileLate` is not clear to me but the dependency looks a little bit dubious. At least we made that dependency explicit.

ScaladocGlobal and Global defined in interactive has been adapted in a way that makes them compile both with quick.comp and 2.11.0-M4 so my refactorings are not blocking the modularization effort.

gkossakowski commented:
I submitted this PR so I can see the test results. I'll assign a reviewer once all tests pass.

gkossakowski committed:
Move ICodeReader-specific logic out of ClassfileParser.

ClassfileParser contained some ICodeReader-specific logic
like `forceMangledName` and `getMemberSymbol` methods.

The `getMemberSymbol` method was defined in ConstantPool class
because it must access some internal state of ConstantPool. In order
to move that method to ICodeReader we had two options:

  1. Make all internal state accessible from outside of ConstantPool
     class so getMemberSymbol could be implemented outside of
     ConstantPool hierarchy
  2. Make it possible to subclass ConstantPool in ICodeReader so
     getMemberSymbol can be implemented in a subclass and can access
     internal state of ConstantPool

Given the fact that getMemberSymbol mutates ConstantPool's internal
state I decided that subclassing is a cleaner approach. It required
significant refactoring because we had to make sure that we create
an instance of proper class when initializing the `pool` variable.
I ended up introducing `ConstantPoolManager` class which is essentially
a mutable variable that can be reset multiple times.

This change makes ClassfileParser independent from the ICodeReader and
its implementation details.

gkossakowski committed:
Fix Platform type in Global to be JavaPlatform.

At the moment we have just one Platform: JavaPlatform in the
compiler. The whole Platform abstraction feels dubious and
Java backends need to downcast to JavaPlatform to implement some
optimizations. It seems like for now it's just better to fix platform
declared in Global to be JavaPlatform and get rid of downcasting.

I checked that even JavaScript backend declares itself as a subtype
of JavaPlatform so it seems like our abstraction is not that useful.

If we have an alternative platform with specific requirements we
might want to refactor our Platform abstraction again but for now
it seems dubious to pay a price for abstraction nobody uses.

gkossakowski committed:
Add example of SymbolTable unit test.

Add a SymbolTableTest which contains all the code needed to
initialize a SymbolTable in JUnit environment. It shows
that initialization of definitions works and one can easily
lookup some symbols and perform tests like subtyping tests.

gkossakowski committed:
Update sbt interface version to 0.12.4.

In particular, that version contains the
sbt/sbt@1b897a3a9422b698bf2c04103169a5eef1ff5b43 commit which removes
all residence compiler infrastructure and makes incremental compiler
less dependent on compiler internals.

gkossakowski committed:
Refactor the cake so SymbolTable does not depend on Global

This is rather large commit so I'll first explain the motivation
behind it and then go through all changes in detail explaining
the choices I made.

The motivation behind this refactoring was to make SymbolTable
unit testable. I wanted a lightweight way of initializing
SymbolTable and then writing unit tests for subtyping algorithm,
various functionality related to Symbols, etc.
All of that should be possible by precisely controlling what we
test, e.g., create types and symbols by hand and not have them
defined in source code as we normally do in partest (functional)
tests.

The other motivation was to reduce and clarify dependencies we
have in the compiler. Explicit dependencies lead to cleaner
design. Also, explicit and reduces dependencies help incremental
compilation which is a big problem for us in compiler's code
base at the moment.

One of the challenges I faced during that refactoring was
cyclic dependency between Platform and SymbolLoaders.

Platform depended on `SymbolLoaders.SymbolLoader` because it
would define a root loader. SymbolLoaders depended on Platform
for numerous reasons like deferring decision how to load a given
symbol based on some Platform-specific hooks.

I decided to break that cycle by removing methods related to
symbol loading from Platform interface. One could argue, that
better fix would be to make SymbolLoaders to not depend on Platform
(backend) concept but that would be much bigger refactoring. Also,
we have a new concept for dealing with symbol loading: Mirrors.

For those reasons both `newClassLoader` and `rootLoader`
were dropped from Platform interface.

Note that JavaPlatform still depends on Global so it can
access phases defined in Global to implement `platformPhases`
method.

Both GenICode and BCodeBodyBuilder have some Platform specific
logic that requires casting because pattern matcher doesn't narrow
types to give them a proper refinement. Check the changes for details.

Some logging utilities has been moved from Global to SymbolTable
because they are accessed by SymbolTable. Since Global inherits from
SymbolTable this should be a source compatible change.

The SymbolLoaders has dependency on `compileLate` method defined in Global.
The purpose behind `compileLate` is not clear to me but the dependency looks
a little bit dubious. At least we made that dependency explicit.

ScaladocGlobal and Global defined in interactive has been adapted in a way
that makes them compile both with quick.comp and 2.11.0-M4 so my refactorings
are not blocking the modularization effort.

gkossakowski committed:
Deprecate Platform.BinaryRepr.

The 27c8266e introduced `Platform.BinaryRepr` abstract type that
was meant to abstract over binary representation of a classfile.

The abstraction was needed at the time because we had both jvm
backend and msil backend. The msil backend has been removed but
the abstraction was kept in place. The type has been deprecated
and set to be an alias for AbstractFile. The alias will be
removed later in 2.11 milestone cycle.

All references to `Platform.BinaryRepr` has been removed in
the compiler in favour of using AbstractFile directly.

If we ever need to abstract over multiple backends we should
do that based on real requirements of given backends.

This is a first step towards breaking the cyclic dependency
between Platform and SymbolLoaders.

gkossakowski committed:
Remove dependency on typer phase in ClassfileParser.

ClassfileParser depends on forcing infos during typer phase.
It's not entirely clear why this is needed (git blame doesn't
help much) but I decided to preserve the logic.

Therefore, I introduced an abstract method called
`lookupMemberAtTyperPhaseIfPossible` which preserves the original
semantics but is implemented outside of ClassfileParser so the
ClassfileParser itself doesn't need to depend on typer phase and
phase mutation utilities.

I removed `loaders` override in `test/files/presentation/doc`.
I would have to update it to implement the
`lookupMemberAtTyperPhaseIfPossible` method. However, the override
doesn't seem to be doing anything useful so I just removed it.

gkossakowski committed:
Make junit runner quiet when tests pass.

Do not show output printed to standard output by default.
When a test fails it will be recorded in a report but when it
passes we don't need any debugging logging to be printed.

gkossakowski committed:
The `test-junit` Eclipse project depends on `scala-compiler`

We want to write JUnit tests for Scala compiler so we should depend
on `scala-compiler` project in `test-junit`.

Also, make dependencies of `scala-compiler` project exported so it's
enough to depend on `scala-compiler` to get all the transitive
dependencies that are needed by the compiler. Otherwise, one would
need to copy all dependencies of Scala compiler to all Eclipse projects
that depend on the compiler.

gkossakowski commented:
I fixed the test failures and addressed most of @retronym's feedback.

Here's a screenshot of JUnit test passing in Eclipse:
![Imgur](http://i.imgur.com/wkJXufF.png)

From there debugging, quick testing is really easy.

retronym commented:
Here's what I had in mind as an alternative to `ConstantPoolManager`: https://github.com/retronym/scala/commit/5172b2a2b.

retronym commented:
And here's what I'd prefer to avoid duplication in lookupMemberAtTyperPhaseIfPossible and compileLate https://github.com/retronym/scala/commit/abf7a46bf2.

Otherwise, this looks really promising.

gkossakowski commented:
@retronym: I cherry picked your changes. I did some renaming because I found `AbsConstantPool` to be confusing name when that class is not abstract.

Also, I couldn't remove duplication from Scaladoc because it has to be _both_ by Scala 2.11.0-M4 and quick. Once 2.11.0-M5 is released we'll be able to clean this up.

gkossakowski commented:
Review by @retronym 

I think it's a good time to assign a reviewer.

gkossakowski commented:
PLS REBUILD/pr-rangepos-per-commit@f3ed70c

scala-jenkins commented:
(kitty-note-to-self: ignore 21676714)
:cat: Roger! Rebuilding pr-rangepos-per-commit for f3ed70c7. :rotating_light:

gkossakowski commented:
@retronym: it seems like beb43f3 will be tricky to get right.

If you are wondering why do we have this weird requirement to be able to build scaladoc with 2.11.0-M4 and quick let me explain. We need to be able to build with quick because that's what we do right now in this repo. However, @adriaanm wants to move scaladoc out of scala main repo and then we'll be using 2.11.0-M4 to build it until M5 is released.

retronym commented:
Perhaps you could submit immediately submit another PR that addresses the "// TODO M5" comments? That PR would be based on this one, but we would delay merging.

gkossakowski commented:
That's a good idea. However, this shows very nicely that evolving APIs based on inheritance is tricky. I think that's the other reason I don't like the cake pattern: it encourages (too) heavy use of inheritance.

gkossakowski commented:
@retronym: the whole compatibility story is actually more complicated, see: https://groups.google.com/d/msg/scala-internals/_abx_dctVQU/YSKyEv6420QJ

gkossakowski commented:
At this point I don't even know how this next PR should look like exactly.

adriaanm commented:
We can always release 2.11.0-M4-1 or 2.11.0-M5-1 with the breaking changes and build a new version of scaladoc against that.

gkossakowski commented:
@adriaanm: Could elaborate on how would you see the whole process? I don't see how everything would fit together. That's why I started the discussion here: https://groups.google.com/d/msg/scala-internals/_abx_dctVQU/YSKyEv6420QJ

adriaanm commented:
Let's assume we have published scala 2.11.0-M4 and scaladoc_2.11.0-M4 1.0, and we want to change Scala's master branch in a way that breaks _compiling_ scaladoc without breaking PR validation and the nightlies on scala and scaladoc. You submit two PRs: one against scala/scala, and the corresponding compensating change for scala/scaladoc. 

At this point, without further work, the PR validation for scaladoc that compiles against 2.11.0-M4 would fail. PR validation for scala will fail to run the published scaladoc_2.11.0-M4 1.0 against the result of compiling the PR. 

A dbuild run against the branches corresponding to both PRs will succeed however (explained below). I'd propose to have a dbuild jenkins job where you can specify the PRs for different dependencies, so that you can try the combination before merging anything. On success, you break the circular dependency by merging the scaladoc PR and doing a rebuild on the scala/scala PR, which should succeed as you're running against scaladoc master (which was tested as the PR by that manual dbuild run). 

As you can still _run_ the published scaladoc_2.11.0-M4 1.0 on the sources in your scala PR, the default dbuild (that uses master of all dependencies) can build and (locally) publish Scala 2.11.0-$sha-SNAPSHOT for PR validation, and compile scala/scaladoc's master (that now has the fix corresponding to the incompatible change) to re-verify that it works. If all is well, we merge the PR and publish Scala 2.11.0-M4-1. Then, we change scaladoc's build to compile against Scala 2.11.0-M4-1. As soon as this version is available, scaladoc's nightly and PR validation will succeed again.

Basically, this recreates the quick/locker bootstrap that we have before by making those steps explicit with publishing new versions. This is tedious and requires some coordination and restarting PR validation, but that's the point: such breaking changes must be rare, and should probably only happen in the milestone cycle. Luckily, we've been automating our release process, and we have dbuild to push this all the way.

gkossakowski commented:
@adriaanm: this makes sense to me. We have discuss details like whether we want to release M4-1 version or we want to rely on SNAPSHOT dependencies for a short period of time. One way or another, I'm confident we'll arrive at some reasonable solution that can accommodate for changes that evolve interaction between Scaladoc and the compiler.

Now, the question is what do we do with this PR? The answer depends on answer to other question: are we still aiming at splitting scaladoc out from scala/scala before M5? The scheduled date for M5 is 12th of August. We have to complete the process of splitting at least 5 days before the release date so we can deal with all potential issues that might arise. Therefore, we should have PR 7-10 days before the release date so it can be reviewed, etc. Therefore we would need PR ready by Friday or Monday next week. Do you think it's realistic deadline?

If we push scaladoc refactoring to M6 then I believe this PR can be merged without any risk.

WDYT?

adriaanm commented:
LGTM. I agree we need to delay the scaladoc refactoring, so this PR is good to go.

gkossakowski commented:
@retronym: are you fine merging it? I'll take care of cleaning TODO comments in a separate PR. It entails changing Eclipse projects definitions and other places where it already assumes that scaladoc has to be built against STARR and not quick.

retronym commented:
Lgtm.  A few rough edges, but they are well explained and/or tracked in tickets.

gkossakowski committed:
Address TODOs around SymbolLoaders and SymbolTable.

SymbolTable refactoring introduced some TODOs that were supposed
to be addressed after M5 release. The reason I couldn't address
those problems right away was a conflict with our plans to modularize
Scaladoc and interactive. However, we decided to delay that work
until after M5 is released so addressing TODOs is not blocked
anymore.

This commit introduces the following changes:

  * Eclipse project definitions for interactive and scaladoc
    depend on scala-compiler project so they are builded against
    latest version of the compiler (quick) instead of STARR.
    This aligns our Eclipse project definitions with build.xml
    structure.
  * Introduce GlobalSymbolLoaders class which wires dependencies
    of SymbolLoaders with assumption of dependency on Global.
  * Switch to GlobalSymbolLoaders in BrowsingLoaders,
    interactive Global and ScaladocGlobal; this eliminates all
    TODO comments introduced before

gkossakowski commented:
@retronym: In reflection, I decided to finish this PR so everybody is happy with it. I want save us from context switching. Let me know if 3cb3c8e addresses all of your concerns. (I'm assuming I didn't break any tests this time).

gkossakowski commented:
The latest failure seems to be coming from PR validation script being broken. I already let @adriaanm know.

adriaanm commented:
PLS REBUILD/pr-rangepos-per-commit@3cb3c8e

scala-jenkins commented:
(kitty-note-to-self: ignore 21841715)
:cat: Roger! Rebuilding pr-rangepos-per-commit for 3cb3c8e3. :rotating_light:

gkossakowski commented:
@adriaanm: What's the status of IDE validation? I'd like to prioritize this PR for testing. I believe it has potential to break IDE in subtle ways because it touches Global, SymbolLoaders and Scaladoc.

adriaanm commented:
https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/542/console failed for d898ca3, not sure why

adriaanm commented:
- 526f6c3 [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/517/console)
- 5eb4cdf [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/528/console)
- afbee09 [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/536/console)
- ced7214 [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/516/console)
- e5121c8 [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/531/console)
- e96962d [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/540/console)
- f3ed70c [ok](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/529/console)

adriaanm commented:
I'll start builds for 115e8b4 and 3cb3c8e3

gkossakowski commented:
> https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/542/console failed for d898ca3, not sure why 

That really looks like a spurious failure. If anything, I'd expect compile-time errors. This commit doesn't change any runtime behavior. Do we have an option to restart validation for that commit?

adriaanm commented:
Sure, I'll restart it as well.

adriaanm commented:
https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala/338/console
https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala/339/console
https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala/340/console

dragos commented on src/compiler/scala/tools/nsc/backend/JavaPlatform.scala:
```diff
@@ -11,33 +11,22 @@ import util.{ClassPath,MergedClassPath,DeltaClassPath}
 
 trait JavaPlatform extends Platform {
+  val global: Global
+  override val symbolTable: global.type = global
```
Why do you need both `global` and `symbolTable`?

gkossakowski replied:
Platform depends just on `SymbolTable`. JavaPlatform depends on `Global` to implement its functionality. Ideally, JavaPlatform would not depend on Global but I decided to leave this for another refactoring round.

dragos replied:
But the two fields have the same type, and the same value. So if anyone knows about `global`, it must know it's using `JavaPlatform`, so it could simply use `symbolTable`.

gkossakowski replied:
I find it confusing that a field called `symbolTable` has a type `Global` (after widening) but that's how we wire our dependencies. The fact that `Global` is subtype of `SymbolTable` is purely accidental in my opinion. Also, I'd like to stop us depending on Global anywhere is source code as I believe it's an anti-pattern but I'll discuss this on internals first.

dragos commented on src/compiler/scala/tools/nsc/backend/icode/ICodes.scala:
```diff
@@ -108,6 +109,12 @@ abstract class ICodes extends AnyRef
+      global.loaders.lookupMemberAtTyperPhaseIfPossible(sym, name)
+    lazy val symbolTable: global.type = global
+    lazy val loaders: global.loaders.type = global.loaders
+    def classPath: util.ClassPath[AbstractFile] = ICodes.this.global.platform.classPath
```
Nitpicking, but do you need the full prefix here? Can't it be just `global.platform.classPath`, or even `platform.classPath`?

gkossakowski replied:
I prefer to be very explicit when I wire dependencies together. It was conscious decision :-)

dragos replied:
So why `import global._` then? :)

gkossakowski replied:
To not create too big change at once. Consider this PR to be a first step of the long process of cleaning up our code base :)

dragos commented on src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala:
```diff
@@ -20,8 +20,23 @@ import scala.reflect.io.{ AbstractFile, NoAbstractFile }
+  def lookupMemberAtTyperPhaseIfPossible(sym: Symbol, name: Name): Symbol
+  /**
+   * Should forward to `Run.compileLate`. The more principled fix would be to
+   * determine why this functionality is needed and extract it into a separate
```
Here's why it's needed: Assume the type-checker needs the info of a symbol that's not been loaded yet. `info` will trigger its symbol loader, and this might be a `source file loader`, meaning that either there is no corresponding classfile for that symbol, or the source is newer.

The source file loader needs to parse and enter the new source, and that's done through `compileLate`. The difference between `compileSources` and `compileLate` is that the first one compiles all sources in lock step: parsing, naming, typing, etc. The `late` version is an out-of-band process to bring a source through all the previous phases until the current one.

This functionality is used when bootstrapping, or when compiling a subset of the project (I think the ant build _might_ rely on it as well). But it's heavily used by the presentation compiler, when in many cases you open sources without first building the project (therefore no classfiles are available). However, this is not specific to the presentation compiler, it's been present since the dawn of scalac.

gkossakowski replied:
Thanks a lot for explanation! :-)

I have some more questions:
1. How is the connection between source file name and symbols it defines established? In particular, what if we are trying to load symbol for `A` but there's no class file corresponding to it and `A` is defined in `B.scala`?
2. What does it mean out-of-band process exactly? Does it mean that we parse the file, then we apply namer/typer to that file and all other subsequent phases? Do we really execute tree transformers from all previous phases until current one?
3. I realize that this mechanism is used to implement `-sourcepath` and thus allows us to compile Scala library. However, I wonder if the logic for this mechanism should live in SymbolLoaders. Or if we could create a separate component solely responsible for this out-of-band process and its semantics. SymbolLoaders could depend on that component.

I would love to hear your feedback. I'd love to understand our process of loading symbols a little bit better.

dragos replied:
> How is the connection between source file name and symbols it defines established? In particular, what if we are trying to load symbol for A but there's no class file corresponding to it and A is defined in B.scala?

In the same way as for classfiles: The package loader enters symbols for each filename, and links the `info` of that symbol with the corresponding loader (a classfile loader, source file loader, or package loader). If the top-level symbol is defined in a file with a different name, it won't be entered ("not found"). This is the crux of the IDE limitation, requiring the folder/filename convention that Java uses.

There is a partial fix: `BrowsingLoaders` are a special loader, swapped as the package loader for the interactive compiler. It does a very fast parse phase of each file in a directory, and enters symbols for all top-level definitions (as opposed to just one symbol per file, with the name of the file). This would solve the `class A` defined in `B.scala`, as long as the folder/package match.

> What does it mean out-of-band process exactly? Does it mean that we parse the file, then we apply namer/typer to that file and all other subsequent phases? Do we really execute tree transformers from all previous phases until current one?

No, it never goes farther than typer. The whole code is pretty short, check `Global.compileLate(CompilationUnit)`. It does the minimal amount of work necessary to recover type signatures from source files.

> I realize that this mechanism is used to implement -sourcepath and thus allows us to compile Scala library. However, I wonder if the logic for this mechanism should live in SymbolLoaders. Or if we could create a separate component solely responsible for this out-of-band process and its semantics. SymbolLoaders could depend on that component.

It is not much more than those 6 lines of code.

gkossakowski replied:
> It is not much more than those 6 lines of code.

The way those 6 lines of code fit into big picture is what I find tricky. You can introduce a lot of complexity in 6 lines of code :)

In any case, thanks for explanation. I think understand symbol loading a little bit better now.

Also, do you know why we are not using BrowsingLoaders by default? It seems like the only reasonable choice gives Scala semantics.

dragos replied:
> > It is not much more than those 6 lines of code.
> > The way those 6 lines of code fit into big picture is what I find tricky. You can introduce a lot of complexity in 6 lines of code :)

I don't think that's the case here. If you have any more questions about how it fits in the big picture, I'm happy to clarify. But I feel that any more indirection (as I don't see any abstraction that would fit in there) would make it more opaque.

> Also, do you know why we are not using BrowsingLoaders by default? It seems like the only reasonable choice gives Scala semantics.

I had the same reaction. I don't know for sure, but I think it's for performance reasons (it has to go through a parsing phase even before "forcing" a symbol -- though it's an `OutlineParser`, meaning it skips all local definitions).

dragos commented on src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala:
```diff
@@ -20,8 +20,23 @@ import scala.reflect.io.{ AbstractFile, NoAbstractFile }
-  val global: Global
-  import global._
+  val symbolTable: symtab.SymbolTable {
+    def settings: Settings
```
I spent a bit of time to realize that this refinement type won't cause any reflective call, and is only needed to narrow down a member type. It would be better to use `override` in this case, even if not required by scalac. I leave below my initial reaction to this code (followed by building scala again with `-feature` :)).

Since this is a refinement type, calls to this member will be reflective. It might introduce subtle slowdowns (symbol loaders are on the hot path).

gkossakowski replied:
I wanted to put override modifier in there for exactly the same reason you suggest but that's what you try to do that:

```
scala> abstract class A { def foo: Object }
defined class A

scala> type ARefined = A { override def foo: String }
<console>:1: error: illegal start of declaration
       type ARefined = A { override def foo: String }
```

I don't know whether this is by design or not.

dragos replied:
I see, thanks for the explanation.

dragos commented on test/junit/scala/tools/nsc/symtab/SymbolTableTest.scala:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  @Test
+  def basicSubTypeCheck = {
```
It would be good to show a test where you create symbols without relying on them being available on the classpath, for instance comparing two existentials.

dragos commented:
Good code, and I know how hard it is to refactor code in such a sensitive (and old) area of the compiler. Good job!

Besides some minor comments, I'd like to see an example of creating symbols by hand, as I don't immediately see how that's achieved.

gkossakowski committed:
Add sample to SymbolTableTest using custom symbols and type.

Add a test which demonstrates how one can create symbols and types
from scratch and perform sub type check using them.

gkossakowski commented:
@dragos: I added a sample which shows how to create symbols and types from scratch for testing in d5e0f72. Let me know if this clarifies things.

gkossakowski commented:
@retronym: I'm still waiting for you to review 3cb3c8e and (now)  d5e0f72.

dragos commented:
@gkossakowski thanks for the examples!

huitseeker commented on build.xml:
```diff
@@ -1538,7 +1538,7 @@ TODO:
   <target name="test.junit" depends="test.junit.comp">
     <stopwatch name="test.junit.timer"/>
     <mkdir dir="${test.junit.classes}"/>
-    <junit fork="yes" haltonfailure="yes" showoutput="yes" printsummary="on">
```
If you're looking at reducing output to the useful, have you looked into using  the 'brief' formatter ? It seems to do just that (look for the  formatter section in https://ant.apache.org/manual/Tasks/junit.html )

gkossakowski replied:
@huitseeker: I believe JUnit is using `brief` formatter by default, right?

huitseeker replied:
I thought it was `plain`. But, default notwithstanding, that `showouput` means you're supressing standard output on purpose, even on failing tests ? Why ?

gkossakowski commented:
@retronym: ping

retronym commented:
LGTM

dragos commented:
LGTM

gkossakowski commented:
Woohoo! Thanks all!

jsuereth commented:
Guys, is the sbt test-interface stuff missing from the pull request validation now?   That used to run and this PR should have been flagged as breaking....

gkossakowski commented:
@jsuereth: I assume you meant compiler interface and not test interface. This PR is not breaking it because we upgraded it to 0.12.4, see 5eb4cdf included in this PR.
